src/api-service/__app__/onefuzzlib/secrets.py

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+
+from typing import Tuple, cast
+from urllib.parse import urlparse
+from uuid import uuid4
+
+from azure.keyvault.secrets import KeyVaultSecret
+from onefuzztypes.models import SecretAddress, SecretData
+from pydantic import BaseModel
+
+from .azure.creds import get_instance_name, get_keyvault_client
+
+
+def save_to_keyvault(secret_data: SecretData) -> None:
+    if isinstance(secret_data.secret, SecretAddress):
+        return
+
+    secret_name = str(uuid4())
+    if isinstance(secret_data.secret, str):
+        secret_value = secret_data.secret
+    elif isinstance(secret_data.secret, BaseModel):
+        secret_value = secret_data.secret.json()
+    else:
+        raise Exception("invalid secret data")
+
+    kv = store_in_keyvault(get_keyvault_address(), secret_name, secret_value)
+    secret_data.secret = SecretAddress(url=kv.id)
```

Comment:
So, when we save `SecretData` to the keyvault, we actually overwrite the secret data with the newly-created address of the secret, throwing away (this copy of) the raw secret.

This helps explain some of what's going on with `SecretData`'s definition.

What if we only allowed the existence of `SecretData` instances that were already saved to the keyvault? Then it would have fewer valid states and be easier to think about. If needed, we could have a `RedactedData` class for something not yet saved, and then this function could return a `SecretData`/`KeyvaultData` after saving. The runtime difference is then encoded in a type distinction.

Reply:
`SecretData` needs to be backward compatible for the existing notification code. So we need to allow secret not yet saved in the keyvault. new classes can use `SecretAddress` as a backing field and never allow secret not saved to keyvault as you suggested